Add method to get all state events in a room

syvb opened the pull request:
It is useful to be able to get all of the state events that have occurred in a room, instead of needing to use `Room::getCurrentState`, which filters based on the event type and state key.

This is useful for implementing Spaces support, where all events of type `m.space.child` are needed, and we don't know their state keys in advance.

KitsuneRal reviewed (changes requested):
Thanks for the contribution, but there's one shortcoming with it: if a room has 10k members, or another kind of heavy state, every `stateEvents()` invitation will lead to copying all of that state into the returned list (fortunately only pointers will be copied but that can be still a lot). If you really want, for whatever reason, to obtain the whole current state, you might better just return a const reference to the internal `currentState` container, in the same way `messageEvents()` returns a const reference to the timeline. On the other hand, if you need all events of a particular type such as `m.space.child` filtering on the original container is more efficient - basically, you could pass the event type parameter to the function so that it could collect all such events for you. I'm fine if you implement either option; but ideally it'd be great to have both.

syvb committed:
Add method to get all state events in a room

It is useful to be able to get all of the state events that have
occured in a room, instead of needing to use Room::getCurrentState,
which filters based on the event type and state key.

syvb commented:
@KitsuneRal I've updated the PR to make `stateEvents` return the inner `QHash`, and `stateEventsOfType` return a list of the state events of a type.

KitsuneRal commented on lib/room.cpp:
```diff
@@ -1293,6 +1313,17 @@ const StateEventBase* Room::getCurrentState(const QString& evtType,
     return d->getCurrentState({ evtType, stateKey });
 }
 
+const QList<const StateEventBase*>*
```
That's not quite what I was looking for: returning a dynamically allocated `QList` will surely lead to people forgetting to `delete` it afterwards, let alone being inefficient since `QList` is [implicitly-shared](https://doc.qt.io/qt-5/implicit-sharing.html) anyway. Also - `QVector` is equally or more efficient compared to `QList` and [should be used by default](https://doc.qt.io/qt-5/qlist.html#details) (in fact, Qt 6 has `QVector` and `QList` referring to the same class, and it's basically Qt 5's `QVector`).
```suggestion
QVector<const StateEventBase*>
```

KitsuneRal commented on lib/room.cpp:
```diff
@@ -1293,6 +1313,17 @@ const StateEventBase* Room::getCurrentState(const QString& evtType,
+    return d->stateEventsOfType(evtType);
+}
+
+const QHash<StateEventKey, const StateEventBase*> Room::stateEvents() const
```
`QHash` is implicily-shared so copying it is cheap - but it could just be a reference, couldn't it? (Assuming that `Q_INVOKABLE` doesn't work, see the next comment.)
Also: `stateEvents()` is not specific enough, since `Room` actually maintains several (two, as of now) sets of state events. The base state is not exposed in the API as yet but it might be in the future - e.g. as a part of the protected API.
```suggestion
const QHash<StateEventKey, const StateEventBase*>& Room::currentState() const
```

KitsuneRal commented on lib/room.h:
```diff
@@ -508,6 +508,19 @@ class Room : public QObject {
+    /*! This method returns all known state events that have occured in
+     * the room, as a mapping from the event type and state key to value.
+     */
+    Q_INVOKABLE const QHash<StateEventKey, const StateEventBase*> stateEvents() const;
```
As much as I'd like it to work, I don't think `Q_INVOKABLE` is very useful here: AFAIK `QHash` is not converted to anything on QML/JavaScript side, you'd have to use `QVariantMap` instead. As conversion to `QVariantMap` is quite straightforward I'd prefer to leave that to clients and just give access to the original hash-map.
```suggestion
    const QHash<StateEventKey, const StateEventBase*>& currentState() const;
```

KitsuneRal commented on lib/room.h:
```diff
@@ -508,6 +508,19 @@ class Room : public QObject {
+    /*! This method returns all known state events that have occured in
+     * the room of the given type.
+     */
+    Q_INVOKABLE const QList<const StateEventBase*>*
```
Unlike the above, `Q_INVOKABLE` is fine here, with `QVector` being converted to a JavaScript array on the QML side (see https://doc.qt.io/qt-5/qtqml-cppintegration-data.html#sequence-type-to-javascript-array) thanks to `const StateEventBase*` being marked with `Q_DECLARE_METATYPE`.
```suggestion
    Q_INVOKABLE QVector<const StateEventBase*>
```

KitsuneRal commented on lib/room.cpp:
```diff
@@ -224,6 +224,26 @@ class Room::Private {
+    {
+        return currentState;
+    }
+
```
This is unneeded as we're already in the internal implementation domain.
```suggestion
```

KitsuneRal commented on lib/room.cpp:
```diff
@@ -1293,6 +1313,17 @@ const StateEventBase* Room::getCurrentState(const QString& evtType,
+
+const QHash<StateEventKey, const StateEventBase*> Room::stateEvents() const
+{
+    return d->stateEvents();
```
```suggestion
    return d->currentState;
```

KitsuneRal commented on lib/room.cpp:
```diff
@@ -224,6 +224,26 @@ class Room::Private {
+        return currentState;
+    }
+
+    const QList<const StateEventBase*>*
+    stateEventsOfType(const QString& evtType) const
```
```suggestion
    QVector<const StateEventBase*> stateEventsOfType(const QString& evtType) const
```

KitsuneRal commented on lib/room.cpp:
```diff
@@ -224,6 +224,26 @@ class Room::Private {
+    const QList<const StateEventBase*>*
+    stateEventsOfType(const QString& evtType) const
+    {
+        QList<const StateEventBase*>* vals = new QList<const StateEventBase*>();
```
```suggestion
        QVector<const StateEventBase*> vals;
```

KitsuneRal commented on lib/room.cpp:
```diff
@@ -224,6 +224,26 @@ class Room::Private {
+    stateEventsOfType(const QString& evtType) const
+    {
+        QList<const StateEventBase*>* vals = new QList<const StateEventBase*>();
+        QHash<StateEventKey, const StateEventBase*>::const_iterator it =
+            currentState.constBegin();
+        for (; it != currentState.constEnd(); ++it) {
+            const StateEventBase* val = it.value();
```
```suggestion
        for (const auto* val: currentState) {
```

KitsuneRal reviewed (changes requested):
Thanks! I have quite a few technical comments to the code but the overall direction is right. Sorry if I missed something - didn't try to compile all the suggestions.

syvb committed:
Add Room::{stateEventsOfType,currentState}

This is useful for implementing Spaces support, where all events of
type `m.space.child` are needed, and we don't know their state keys in
advance.

KitsuneRal reviewed (approved):
Thanks!
